File: backend/src/models/Resume.js
import mongoose from 'mongoose';

const sectionReviewSchema = new mongoose.Schema({
  section: { type: String, required: true, trim: true, maxlength: 80 },
  status: { type: String, enum: ['strong', 'needs_work', 'missing'], required: true },
  score: { type: Number, required: true, min: 0, max: 100 },
  feedback: { type: String, required: true, trim: true, maxlength: 500 },
}, { _id: false });

const resumeSchema = new mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    fileName: { type: String, required: true, trim: true, maxlength: 160 },
    sourceType: { type: String, enum: ['pdf', 'text', 'pasted'], required: true },
    targetRole: { type: String, default: '', trim: true, maxlength: 120 },
    wordCount: { type: Number, required: true, min: 1, max: 20000 },
    analysisSource: { type: String, enum: ['ai', 'rule_based'], default: 'ai', required: true },
    notice: { type: String, default: '', trim: true, maxlength: 300 },
    analysis: {
      overallScore: { type: Number, required: true, min: 0, max: 100 },
      summary: { type: String, required: true, trim: true, maxlength: 1200 },
      strengths: { type: [{ type: String, trim: true, maxlength: 300 }], default: [], validate: [(items) => items.length <= 8, 'A maximum of 8 strengths is allowed'] },
      improvements: { type: [{ type: String, trim: true, maxlength: 400 }], default: [], validate: [(items) => items.length <= 10, 'A maximum of 10 improvements is allowed'] },
      missingKeywords: { type: [{ type: String, trim: true, maxlength: 80 }], default: [], validate: [(items) => items.length <= 20, 'A maximum of 20 keywords is allowed'] },
      sectionReviews: { type: [sectionReviewSchema], default: [], validate: [(items) => items.length <= 12, 'A maximum of 12 section reviews is allowed'] },
      actionPlan: { type: [{ type: String, trim: true, maxlength: 400 }], default: [], validate: [(items) => items.length <= 8, 'A maximum of 8 action items is allowed'] },
    },
    analyzedAt: { type: Date, default: Date.now, required: true },
  },
  { timestamps: true }
);

resumeSchema.index({ user: 1, analyzedAt: -1 });

const Resume = mongoose.model('Resume', resumeSchema);

export default Resume;

File: ai-service/app/schemas/resume.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

StrengthText = Annotated[str, Field(min_length=1, max_length=300)]
ImprovementText = Annotated[str, Field(min_length=1, max_length=400)]
KeywordText = Annotated[str, Field(min_length=1, max_length=80)]
ActionPlanText = Annotated[str, Field(min_length=1, max_length=400)]


class ResumeAnalysisRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    resumeText: str = Field(min_length=100, max_length=20000)
    targetRole: str = Field(default="", max_length=120)

    @field_validator("resumeText")
    @classmethod
    def validate_resume_text(cls, value: str) -> str:
        cleaned = value.strip()
        if len(cleaned) < 100:
            raise ValueError("Resume text must contain at least 100 characters.")
        return cleaned

    @field_validator("targetRole")
    @classmethod
    def normalize_target_role(cls, value: str) -> str:
        return value.strip()


class ResumeSectionReview(BaseModel):
    model_config = ConfigDict(extra="forbid")

    section: str = Field(min_length=2, max_length=80)
    status: str = Field(pattern="^(strong|needs_work|missing)$")
    score: int = Field(ge=0, le=100)
    feedback: str = Field(min_length=5, max_length=500)


class ResumeAnalysisResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    overallScore: int = Field(ge=0, le=100)
    summary: str = Field(min_length=20, max_length=1200)
    strengths: list[StrengthText] = Field(default_factory=list, max_length=8)
    improvements: list[ImprovementText] = Field(min_length=1, max_length=10)
    missingKeywords: list[KeywordText] = Field(default_factory=list, max_length=20)
    sectionReviews: list[ResumeSectionReview] = Field(min_length=1, max_length=12)
    actionPlan: list[ActionPlanText] = Field(min_length=1, max_length=8)


class ResumeAnalysisEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid")

    analysis: ResumeAnalysisResponse
    wordCount: int = Field(ge=1, le=20000)
    analysisSource: Literal["ai", "rule_based"] = "ai"
    notice: str = Field(default="", max_length=300)
